### ISSUE TITLE
plainToClass assign unknown properties

### ISSUE THREAD

**Issue by nimaen** [typestack/class-transformer#198]
Hi,
I don't know if this is a known issue since I could not find any.
Fact is, the `plainToClass` method assigns unknown properties to the class.

```typescript
class User {
  id: number
  firstName: string
  lastName: string
  password: string
}

const fromPlainUser = {
  unkownProp: 'hello there',
  firstName: 'Umed',
  lastName: 'Khudoiberdiev',
  password: 'imnosuperman'
}

console.log(plainToClass(User, fromPlainUser))
```

Result is :
```
User {
  unkownProp: 'hello there',
  firstName: 'Umed',
  lastName: 'Khudoiberdiev',
  password: 'imnosuperman' 
}
```

The only workaround that I saw looking at the doc would be to add an `@Exclude` decorator on the class and an `@Expose` decorator on each property, which is not really convenient...
Any other workaround ? Is this a known thing ?

**Comment by NoNameProvided** [comment]
Yes, it is known and it will be solved in the same way as in the `class-transformer` lib. You will be able to provide an extra parameter to strip unknown properties.

**Comment by nimaen** [comment]
I don't understand, this is the `class-transformer` lib 🤔

**Comment by NoNameProvided** [comment]
Sorry, updated my comment, I meant `class-validator`.

**Comment by nimaen** [comment]
Got it 😉 
I somehow inferred something like that.
Should I do a PR for this then ?

**Comment by henrikra** [comment]
@nimaen Please do :)

**Comment by quen2404** [comment]
Any news about this ?

**Comment by nimaen** [comment]
Sorry, I somehow forgot about it 👼 
I'll try to have a look in the week.

**Comment by nimaen** [comment]
Ok, so, I've digged down into the reflective api and what we can do with it, and it seems like, per nature, one cannot list a class properties from the constructor metadata **without adding a decorator on each class property**.

We could also list properties names **only** with `Object.getOwnPropertyNames` but it would require to have default values on each property, which is not good enough and does not cover all cases.

So, I'm back at my first comment again : the only way to do it _in a proper way_ and without adding new requirements would be to use the existing `@Expose` decorator, and to add a new option to only assign exposed properties, as @NoNameProvided suggested [here](https://github.com/typestack/class-transformer/issues/198#issuecomment-435694370).

Would that still be ok ?

**Comment by quen2404** [comment]
Yes, i'm agree with @NoNameProvided comment :smile:

**Comment by nimaen** [comment]
Well, there won't be any magic.
It would be only usable like this:

```typescript
class User {
  @Expose() id: number;
  @Expose() firstName: string;
  @Expose() lastName: string;
}

const plainUser = {
  firstName: "Umed",
  lastName: "Khudoiberdiev",
  age: 12
};

plainToClass(User, plainUser, { excludeExtraneousValues: true })
```

And this would log :

```
User (
  id: undefined,
  firstName: "Umed",
  lastName: "Khudoiberdiev"
)
```

**Comment by nguyenbathanh** [comment]
@nimaen Thanks!

I have looked at your previous comment. Seems like there is no approach to adapt with expected behavior in your issue description with current package API (I also checked the source code as well).

**Comment by AmirTugi** [comment]
`excludeExtraneousValues` is not yet included in an npm version :/

**Comment by nimaen** [comment]
Could a collaborator publish a new version please ?
@NoNameProvided maybe ?

**Comment by MTschannett** [comment]
@AmirTugi I'm sorry bitte I cannot create a new Version in npm. @NoNameProvided should be able to so this

**Comment by mitkoevoets** [comment]
This is awesome. Waiting for it to be available in npm version...

**Comment by mitkoevoets** [comment]
This is great. Waiting for it to be available in npm version...

**Comment by mitkoevoets** [comment]
This is desirable. Waiting for it to be available in npm version...

**Comment by MTschannett** [comment]
Yeah i know . I'll try to talk to @NoNameProvided to talk about the right permissions

**Comment by johndoe2361289638165** [comment]
Please release a new version. This option would be so fucking important...

**Comment by unclemu** [comment]
@MTschannett  @NoNameProvided Hey, can we get the latest changes in NPM please, thanks!

**Comment by yesworld** [comment]
I faced same problem! :sob: 
Please release a new version.

**Comment by xorik** [comment]
You could use `{whitelist: true}` on the class-validator instead of `excludeExtraneousValues` in class-transformer

**Comment by pablosproject** [comment]
In the same position here. I user plain class-transformer to convert my DTO objects, and I really need the `excludeExtraneousValues`. Anyone that is available to update it on npm?

**Comment by xorik** [comment]
@pablosproject it's not available in npm. I've tried to build it, but without any success.

But if you use https://github.com/typestack/class-validator for your DTOs, you could set flag `whitelist` to true, it will do the same as `excludeExtraneousValues` for class-transformer

**Comment by lexeek** [comment]
Still waiting for new version with 'excludeExtraneousValues' option included.

**Comment by kunokdev** [comment]
Should we fork this lib ?

**Comment by ullalaaron** [comment]
Hello! Why is this issue closed?

**Comment by nimaen** [comment]
Because I've fixed it in a #222

**Comment by kunokdev** [comment]
@nimaen could you also publish it to npm?

**Comment by nimaen** [comment]
Nah, I'm a mere OSS contributor, I'm not a Typestack member.

**Comment by signal-intrusion** [comment]
waiting....

**Comment by NoNameProvided** [comment]
It's released in `0.2.1`.

**Comment by ljobse** [comment]
For people who are using this in combination with class-validator and thus are decorating the objects with @IsString() for example anyway. 

I wrote a little extension that uses those decorators to strip all unknown properties from the object.
https://gist.github.com/ljobse/a617ca62e7d0277a8da351479656f04f

The file just needs to be loaded and the plainToClass function will now strip all properties which are not decorated with anything.
